test(networking): speed up and simplify address sim

joshuef opened the pull request:
This should open the door to the sim being muuuuch larger. 

goes from 65s on main-> 0.1s on my machine.... 

Could well be I've messed up somethiing in the sim to get that... But it appears to get similar output to pain (still fails a lot with same params, i'd increase the iterations or counts eg...



This pull request introduces significant improvements and refactoring to the `sn_networking/src/record_store.rs` file, particularly within the test module. The changes focus on simplifying the code, improving performance, and enhancing readability. The most important changes include removing unused imports, refactoring the `address_distribution_sim` test, and updating the `pick_cheapest_payee` function.

### Code Simplification and Cleanup:
* Removed unused imports such as `sort_peers_by_key` and `BTreeMap` from the test module (`sn_networking/src/record_store.rs`).

### Refactoring:
* Introduced the `PeerStats` struct to encapsulate peer statistics, replacing the previous `HashMap` approach. This improves code readability and maintainability (`sn_networking/src/record_store.rs`).
* Refactored the `address_distribution_sim` test to use the `PeerStats` struct and `rayon` for parallel iteration, enhancing performance and clarity (`sn_networking/src/record_store.rs`).
* Updated the `pick_cheapest_payee` function to work with the new `PeerStats` struct and use a more efficient approach to find the cheapest payee (`sn_networking/src/record_store.rs`).

### Performance Improvements:
* Utilized `rayon::prelude::*` for parallel iteration in the `address_distribution_sim` test, significantly improving the performance of the simulation (`sn_networking/src/record_store.rs`).
* Optimized the sorting and partitioning logic within the `address_distribution_sim` test to efficiently find and sort peers based on their distance to a target address (`sn_networking/src/record_store.rs`).

These changes collectively enhance the codebase by making it more efficient, easier to understand, and maintain.

bochaco commented:
bors try

joshuef committed:
test(networking): speed up and simplify address sim

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -1461,120 +1460,140 @@ mod tests {
-            let _ = peers.insert(peer_id, (0, 0, 0));
-            peers_vec.push(peer_id);
-        }
+        let close_group_size = close_group_majority().min(num_of_peers);
```
I think this shall be replication_range to be used.
in practical, this shall be a kbucket_index
in the sim, we used REPLICATION_PEERS_COUNT (7) to simulate, which shall actually be updated to (K_VALUE/2) to be more close to the real situation.
but change to majority of close_group_size (which is just 3) will be too aggressive

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -1461,120 +1460,140 @@ mod tests {
+                    let mut left = partition_point;
+                    let mut right = partition_point;
+
+                    while close_group.len() < close_group_size && (left > 0 || right < peers_len) {
```
the `partition_point` returns the first index that not satisfy the judgement

using this sorted list against a random target, to figure out closest peers against another target,
the current code block has following glitches:
1, the `left` side shall be preferred, as their `distance interval` is smaller than the `right` side peers
2, the distance from different observer will be different, i.e. the peers around the partition_point not guaranteered to always being the closest peers to the new target
3, the while loop will terminate earlier when the partition_point  falls into the beginning ([0,1,2]) or end ([len-2, len-1, len]) of the entire peers list.

for the point 1 and 2, the `glitch` might be acceptable, as the test is just a `sim`

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -1461,120 +1460,140 @@ mod tests {
-                        panic!("Can't find replicate range of {name:?} with error {err:?}")
+
+                    // Sort the close group by distance to the chunk
+                    close_group.sort_by_key(|&i| chunk_address.distance(&peers[i].address));
```
there is `while close_group.len() < close_group_size` in the above
which means this sort and truncate is not necessary.

unless you mean the previous loop shall use something like `while close_group.len() < 5 * close_group_size` to reduce the chance of glitch that described above ?

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -1461,120 +1460,140 @@ mod tests {
+                .collect();
+
+            // Update peers stats
+            for (peers_in_replicate_range, payee_index) in chunk_results {
```
the `pick_cheapest_payee` uses the payment_received & chunks_stored to calculate cost and payee
which means the peer_status shall be updated each time when a chunk updated

hence, I think this `peer stats update block` shall be inside the above while loop, not outside

joshuef replied:
Okay, moved it into a sequential loop after detecting the desired peers just now 👍

joshuef committed:
feat(networking): simplify the store cost, remove livetime + payment

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -1461,199 +1412,205 @@ mod tests {
-        } else {
-            panic!("Cannot find cheapest payee among {peers_in_close:?}");
-        }
+    fn pick_cheapest_payee(peers: &[PeerStats], close_group: &[usize]) -> usize {
```
ideally, shall use the same `network::get_fees_from_store_cost_responses` function

joshuef replied:
hmm, good shout. This algo is wayyy less fair than just choosing lowest cost + no tiebreak atm... 

I guess as we pass in sorted peers (from the test), so the lowest index is effectively random on the chunk store addr. 

But then when we compare addresses, we actually end up always choosing the same... 

Hmmm. If the algo is just "random", that may actually be fine (i mean here, we dont rely on the order passed in, but not doing anything deterministic because we dont need to?)

joshuef committed:
test: use get_fees_from_store_cost_responses for address_sim pricing

joshuef committed:
test: cleanup address_sim

joshuef committed:
chore(networking): simplify fee finding

Sorting prices by address as fallback is overly deterministic and
tends to skew payeers (as shown by address sim).

Now we add in randomness instead and get a much wider coverage of
paid nodes, and we can see a much faster pull towards similar
earnings over time for nodes
